### PR TITLE
chore: remove stats batcher and workspace usage tracker

### PR DESCRIPTION
**Pull request by f0ssel**
It's become clear that only `workspacestats.ReportAgentStat()` is the only caller of `StatsBatcher.Add()`. Using the stats batcher saves us a DB call, but only 1 call is saved and many are still called outside of the batcher for reporting a stat. I don't think the StatsBatcher caching abstraction is serving us anymore, so I think it should be removed in favor of writing directly to the DB when called.

### CODE REVIEW

File: coderd/workspacestats/reporter.go [hunk 1]
```diff
@@ -146,11 +146,46 @@ func (r *Reporter) ReportAgentStats(ctx context.Context, now time.Time, workspac
...
+			r.opts.Logger.Error(ctx, "unable to marshal agent connections by proto, dropping data", slog.Error(err))
+		} else {
+			connectionsByProto = payload
 		}
+
+		params := database.InsertWorkspaceAgentStatsParams{
+			ID:                          []uuid.UUID{uuid.New()},
+			CreatedAt:                   []time.Time{dbtime.Time(now)},
+			AgentID:                     []uuid.UUID{workspaceAgent.ID},
+			ConnectionsByProto:          connectionsByProto,
```
Comment (johnstcn) [inline coderd/workspacestats/reporter.go:162]: I think this needs to be an array

**Comment by mafredri** [comment]
I was under the impression that `batchstats` buffers up to a certain amount of updates, so let's say 100 workspaces all report their agent stats at the same time, all 100 stats would be written in one go. Is this not the case? (Please note, I'm not making a case for or against the batcher here, just wondering if the premise is correct?)

**Comment by johnstcn** [comment]
> I was under the impression that `batchstats` buffers up to a certain amount of updates, so let's say 100 workspaces all report their agent stats at the same time, all 100 stats would be written in one go. Is this not the case? (Please note, I'm not making a case for or against the batcher here, just wondering if the premise is correct?)

Yep, up to 1024 stats updates.

**Review by johnstcn — DISMISSED** [review]
I'm happy to remove this if it's not enough of a net positive!

**Comment by f0ssel** [comment]
The batcher is totally valid, but my concern is that we are only saving 1 db call when reporting the stat already takes 7-8 db calls to get to that point, and continues to use the DB afterwards too. In return we lose the ability to return an error on failed writes, and a good bit of code to manage this buffer.

**Comment by f0ssel** [comment]
Going to continue removing some batchers on this branch and going to scaletest the impact to see if these changes are safe to merge

**Comment by mafredri** [comment]
> The batcher is totally valid, but my concern is that we are only saving 1 db call when reporting the stat already takes 7-8 db calls to get to that point, and continues to use the DB afterwards too. In return we lose the ability to return an error on failed writes, and a good bit of code to manage this buffer.

I see, that's fair reasoning. I believe writing to this table (agent stats) used to be more expensive as we kept 6 months of data in it. Now we only keep about 1 day.

**Comment by f0ssel** [comment]
@mafredri Thanks for bringing this up, I do share your concerns since it's a write operation, so I've went ahead and made an alternate PR where we keep these structures in place under the `workspacestats.Reporter`. I'd also like your perspective on scale testing: If we do a scale test and find that these batchers are not showing any impact (that we measure), would you still want to keep them for posterity or would you be in favor of removing them?

**Comment by mafredri** [comment]
> @mafredri Thanks for bringing this up, I do share your concerns since it's a write operation, so I've went ahead and made an alternate PR (#13418) where we keep these structures in place under the `workspacestats.Reporter`.

👍🏻

> I'd also like your perspective on scale testing: If we do a scale test and find that these batchers are not showing any impact (that we measure), would you still want to keep them for posterity or would you be in favor of removing them?

If we find that they're not beneficial, I'm fine with removing them entirely. We can always dig in the git history if we realize we need them back.

File: coderd/workspacestats/reporter.go [hunk 2]
```diff
@@ -146,11 +147,45 @@ func (r *Reporter) ReportAgentStats(ctx context.Context, now time.Time, workspac
 
 	var errGroup errgroup.Group
 	errGroup.Go(func() error {
-		err := r.opts.StatsBatcher.Add(now, workspaceAgent.ID, workspace.TemplateID, workspace.OwnerID, workspace.ID, stats)
+		start := time.Now()
+		connectionsByProto := json.RawMessage(`[]`)
```
Comment (mafredri) [inline coderd/workspacestats/reporter.go:151]: I wonder if `[{}]` would be a more accurate representation here? 🤔 Not sure it's necessary though.

**Review by mafredri — APPROVED** [review]

**Review by johnstcn — COMMENTED** [review]
After performing a 1,000 user scaletest and comparing with results from yesterday. Given a GCP `db2-custom-8` instance with 100GB of disk:

- Before: 
  - With no active connections, seeing DB CPU usage of around 80% on a GCP `db2-custom-8` instance.
  - With active SSH connections, seeing DB CPU usage of around 90%
  - InsertWorkspaceAgentStats averaging 1.2 q/s
- After:
  - With no active connections, seeing DB CPU usage of around 86%
  - With active SSH connections, seeing DB CPU usage maxing out.
  - InsertWorkspaceAgentStats averaging 33.3 q/s

The increased queries during active connections is likely due to the agents only reporting stats if there are active connections.

I can't be sure of the wider ramifications of this change, but it does definitely cause increased database load in a scaletest environment. I would recommend against merging this.

**Comment by f0ssel** [comment]
Thanks for this data, glad to close and happy to know these are working as intended!

**Pull request closed without merging**